In [1]:
# This is to install necessary components to run the assignment
# Note: For compatability purposes, libraries have been updated to match to current versions; 
# hence some of the package invocations may differ slightly from the book
!pip install -r requirements.txt

ERROR: Could not find a version that satisfies the requirement numpy==1.26 (from versions: 1.3.0, 1.4.1, 1.5.0, 1.5.1, 1.6.0, 1.6.1, 1.6.2, 1.7.0, 1.7.1, 1.7.2, 1.8.0, 1.8.1, 1.8.2, 1.9.0, 1.9.1, 1.9.2, 1.9.3, 1.10.0.post2, 1.10.1, 1.10.2, 1.10.4, 1.11.0, 1.11.1, 1.11.2, 1.11.3, 1.12.0, 1.12.1, 1.13.0, 1.13.1, 1.13.3, 1.14.0, 1.14.1, 1.14.2, 1.14.3, 1.14.4, 1.14.5, 1.14.6, 1.15.0, 1.15.1, 1.15.2, 1.15.3, 1.15.4, 1.16.0, 1.16.1, 1.16.2, 1.16.3, 1.16.4, 1.16.5, 1.16.6, 1.17.0, 1.17.1, 1.17.2, 1.17.3, 1.17.4, 1.17.5, 1.18.0, 1.18.1, 1.18.2, 1.18.3, 1.18.4, 1.18.5, 1.19.0, 1.19.1, 1.19.2, 1.19.3, 1.19.4, 1.19.5)
ERROR: No matching distribution found for numpy==1.26


In [14]:
from __future__ import print_function
import os
import warnings
# Suppress TensorFlow WARNING logs
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '2'
# Suppress Python Future Warnings
warnings.simplefilter(action='ignore', category=FutureWarning)
import numpy as np
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.optimizers import SGD
from tensorflow.keras.utils import to_categorical

In [15]:
np.random.seed(1671) #for reproducibility

In [16]:
# Network and Training Variables
NB_EPOCH = 20
BATCH_SIZE = 128
VERBOSE = 1
NB_CLASSES = 10 # number of outputs = number of digits
OPTIMIZER = SGD() # optimizer, explained later in this chapter
N_HIDDEN = 128
VALIDATION_SPLIT = 0.2 # how much TRAIN is reserved for VALIDATION

In [17]:
# data: shuffled and split between train and test sets
(X_train, y_train), (X_test, y_test) = mnist.load_data()


In [18]:
#X_train is 60000 rows of 28x28 values --> reshaped in 60000 x 784

RESHAPED = 784

X_train = X_train.reshape(60000, RESHAPED)
X_test = X_test.reshape(10000, RESHAPED)
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')

In [19]:
# normalize
X_train /= 255
X_test /= 255

print(X_train.shape[0], 'train samples')
print(X_test.shape[0], 'test samples')

60000 train samples
10000 test samples


In [20]:
# convert class vectors to binary class matrices

# Y_train = np_utils.to_categorical() replaced

Y_train = to_categorical(y_train, NB_CLASSES) # Updated to fit current version
Y_test = to_categorical(y_test, NB_CLASSES)

In [21]:
# M_HIDDEN hidden layers
# 10 outputs
# final stage is softmax

model = Sequential()
model.add(Dense(N_HIDDEN, input_shape=(RESHAPED,), activation='relu'))  # First hidden layer
model.add(Dense(N_HIDDEN, activation='relu'))                            # Second hidden layer
model.add(Dense(NB_CLASSES, activation='softmax'))                       # Output layer
model.summary()
model.compile(loss='categorical_crossentropy', optimizer=OPTIMIZER, metrics=['accuracy'])


history = model.fit(X_train, Y_train, batch_size=BATCH_SIZE, epochs=NB_EPOCH, verbose=VERBOSE, validation_split=VALIDATION_SPLIT)

score = model.evaluate(X_test, Y_test, verbose=VERBOSE)

print("Test Score:", score[0])
print("Test Accuracy:", score[1])

Model: "sequential_4"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_12 (Dense)             (None, 128)               100480    
_________________________________________________________________
dense_13 (Dense)             (None, 128)               16512     
_________________________________________________________________
dense_14 (Dense)             (None, 10)                1290      
Total params: 118,282
Trainable params: 118,282
Non-trainable params: 0
_________________________________________________________________
Train on 48000 samples, validate on 12000 samples
Epoch 1/20
48000/48000 [==============================] - 1s 13us/sample - loss: 1.5197 - acc: 0.6090 - val_loss: 0.7636 - val_acc: 0.8401
Epoch 2/20
48000/48000 [==============================] - 1s 12us/sample - loss: 0.6028 - acc: 0.8529 - val_loss: 0.4598 - val_acc: 0.8825
Epoch 3/20
48000/48000 [==========================

In [22]:
#test 1 N Hidden parameter decreased to 8

np.random.seed(1671) #for reproducibility

# Network and Training Variables
NB_EPOCH = 20 
BATCH_SIZE = 128
VERBOSE = 1
NB_CLASSES = 10 # number of outputs = number of digits
OPTIMIZER = SGD() # optimizer, explained later in this chapter
N_HIDDEN = 8 # updated to 8
VALIDATION_SPLIT = 0.2 # how much TRAIN is reserved for VALIDATION
# data: shuffled and split between train and test sets
(X_train, y_train), (X_test, y_test) = mnist.load_data()

#X_train is 60000 rows of 28x28 values --> reshaped in 60000 x 784

RESHAPED = 784

X_train = X_train.reshape(60000, RESHAPED)
X_test = X_test.reshape(10000, RESHAPED)
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
# normalize
X_train /= 255
X_test /= 255

print(X_train.shape[0], 'train samples')
print(X_test.shape[0], 'test samples')

# convert class vectors to binary class matrices

# Y_train = np_utils.to_categorical() replaced

Y_train = to_categorical(y_train, NB_CLASSES) # Updated to fit current version
Y_test = to_categorical(y_test, NB_CLASSES)


model = Sequential()
model.add(Dense(N_HIDDEN, input_shape=(RESHAPED,), activation='relu'))  # First hidden layer
model.add(Dense(N_HIDDEN, activation='relu'))                            # Second hidden layer
model.add(Dense(NB_CLASSES, activation='softmax'))                       # Output layer
model.summary()
model.compile(loss='categorical_crossentropy', optimizer=OPTIMIZER, metrics=['accuracy'])

history = model.fit(X_train, Y_train, batch_size=BATCH_SIZE, epochs=NB_EPOCH, verbose=VERBOSE, validation_split=VALIDATION_SPLIT)

score = model.evaluate(X_test, Y_test, verbose=VERBOSE)

print("Test Score:", score[0])
print("Test Accuracy:", score[1])


60000 train samples
10000 test samples
Model: "sequential_5"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_15 (Dense)             (None, 8)                 6280      
_________________________________________________________________
dense_16 (Dense)             (None, 8)                 72        
_________________________________________________________________
dense_17 (Dense)             (None, 10)                90        
Total params: 6,442
Trainable params: 6,442
Non-trainable params: 0
_________________________________________________________________
Train on 48000 samples, validate on 12000 samples
Epoch 1/20
48000/48000 [==============================] - 0s 8us/sample - loss: 2.1633 - acc: 0.2159 - val_loss: 1.9465 - val_acc: 0.3072
Epoch 2/20
48000/48000 [==============================] - 0s 7us/sample - loss: 1.6738 - acc: 0.4645 - val_loss: 1.3836 - val_acc: 0.5695
Epoch 3/20
48000/

In [23]:
#test 2 N hidden parameter increased to 512

np.random.seed(1671) #for reproducibility

# Network and Training Variables
NB_EPOCH = 20 
BATCH_SIZE = 128 
VERBOSE = 1
NB_CLASSES = 10 # number of outputs = number of digits
OPTIMIZER = SGD() # optimizer, explained later in this chapter
N_HIDDEN = 512 # update to 512
VALIDATION_SPLIT = 0.2 # how much TRAIN is reserved for VALIDATION
# data: shuffled and split between train and test sets
(X_train, y_train), (X_test, y_test) = mnist.load_data()

#X_train is 60000 rows of 28x28 values --> reshaped in 60000 x 784

RESHAPED = 784

X_train = X_train.reshape(60000, RESHAPED)
X_test = X_test.reshape(10000, RESHAPED)
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
# normalize
X_train /= 255
X_test /= 255

print(X_train.shape[0], 'train samples')
print(X_test.shape[0], 'test samples')

# convert class vectors to binary class matrices

# Y_train = np_utils.to_categorical() replaced

Y_train = to_categorical(y_train, NB_CLASSES) # Updated to fit current version
Y_test = to_categorical(y_test, NB_CLASSES)


model = Sequential()
model.add(Dense(N_HIDDEN, input_shape=(RESHAPED,), activation='relu'))  # First hidden layer
model.add(Dense(N_HIDDEN, activation='relu'))                            # Second hidden layer
model.add(Dense(NB_CLASSES, activation='softmax'))                       # Output layer
model.summary()
model.compile(loss='categorical_crossentropy', optimizer=OPTIMIZER, metrics=['accuracy'])

history = model.fit(X_train, Y_train, batch_size=BATCH_SIZE, epochs=NB_EPOCH, verbose=VERBOSE, validation_split=VALIDATION_SPLIT)

score = model.evaluate(X_test, Y_test, verbose=VERBOSE)

print("Test Score:", score[0])
print("Test Accuracy:", score[1])


60000 train samples
10000 test samples
Model: "sequential_6"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_18 (Dense)             (None, 512)               401920    
_________________________________________________________________
dense_19 (Dense)             (None, 512)               262656    
_________________________________________________________________
dense_20 (Dense)             (None, 10)                5130      
Total params: 669,706
Trainable params: 669,706
Non-trainable params: 0
_________________________________________________________________
Train on 48000 samples, validate on 12000 samples
Epoch 1/20
48000/48000 [==============================] - 2s 34us/sample - loss: 1.2610 - acc: 0.7141 - val_loss: 0.6167 - val_acc: 0.8629
Epoch 2/20
48000/48000 [==============================] - 2s 33us/sample - loss: 0.5208 - acc: 0.8703 - val_loss: 0.4100 - val_acc: 0.8933
Epoch 3/20


In [24]:
#test 2 N hidden parameter increased to 1024

np.random.seed(1671) #for reproducibility

# Network and Training Variables
NB_EPOCH = 20 
BATCH_SIZE = 128 
VERBOSE = 1
NB_CLASSES = 10 # number of outputs = number of digits
OPTIMIZER = SGD() # optimizer, explained later in this chapter
N_HIDDEN = 1024 # update to 1024
VALIDATION_SPLIT = 0.2 # how much TRAIN is reserved for VALIDATION
# data: shuffled and split between train and test sets
(X_train, y_train), (X_test, y_test) = mnist.load_data()

#X_train is 60000 rows of 28x28 values --> reshaped in 60000 x 784

RESHAPED = 784

X_train = X_train.reshape(60000, RESHAPED)
X_test = X_test.reshape(10000, RESHAPED)
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
# normalize
X_train /= 255
X_test /= 255

print(X_train.shape[0], 'train samples')
print(X_test.shape[0], 'test samples')

# convert class vectors to binary class matrices

# Y_train = np_utils.to_categorical() replaced

Y_train = to_categorical(y_train, NB_CLASSES) # Updated to fit current version
Y_test = to_categorical(y_test, NB_CLASSES)


model = Sequential()
model.add(Dense(N_HIDDEN, input_shape=(RESHAPED,), activation='relu'))  # First hidden layer
model.add(Dense(N_HIDDEN, activation='relu'))                            # Second hidden layer
model.add(Dense(NB_CLASSES, activation='softmax'))                       # Output layer
model.summary()
model.compile(loss='categorical_crossentropy', optimizer=OPTIMIZER, metrics=['accuracy'])

history = model.fit(X_train, Y_train, batch_size=BATCH_SIZE, epochs=NB_EPOCH, verbose=VERBOSE, validation_split=VALIDATION_SPLIT)

score = model.evaluate(X_test, Y_test, verbose=VERBOSE)

print("Test Score:", score[0])
print("Test Accuracy:", score[1])


60000 train samples
10000 test samples
Model: "sequential_7"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_21 (Dense)             (None, 1024)              803840    
_________________________________________________________________
dense_22 (Dense)             (None, 1024)              1049600   
_________________________________________________________________
dense_23 (Dense)             (None, 10)                10250     
Total params: 1,863,690
Trainable params: 1,863,690
Non-trainable params: 0
_________________________________________________________________
Train on 48000 samples, validate on 12000 samples
Epoch 1/20
48000/48000 [==============================] - 5s 103us/sample - loss: 1.1700 - acc: 0.7572 - val_loss: 0.5606 - val_acc: 0.8777
Epoch 2/20
48000/48000 [==============================] - 5s 101us/sample - loss: 0.4819 - acc: 0.8805 - val_loss: 0.3854 - val_acc: 0.8992
Epoch

| N_HIDDEN | Test Loss | Test Accuracy |Validation Accuracy 
|----------|-----------|---------------|-------------------|
| 8        | 0.319     | 0.909         | 0.909             |
| 128      | 0.233     | 0.933         | 0.934             |
| 512      | 0.221     | 0.939         | 0.937             |
| 1024     | 0.216     | 0.941         | 0.941             |

# Summary
I conducted three tests to examine the N_HIDDEN variable at 8, 512, and 1024. The N_HIDDEN variable represents the number of processing neurons in a hidden layer. When decreasing N_HIDDEN to 8, both the test score and accuracy decreased compared to the original setting. The test score represents the categorical crossentropy, or how incorrect the model’s predictions are, while the accuracy indicates how well the model predicts correctly. By examining the data, we see that as N_HIDDEN decreases, the accuracy drops and the test loss increases. The validation accuracy also decreases, which makes sense because fewer neurons per layer mean the model has less capacity to capture patterns in the data.

The opposite trend was observed with N_HIDDEN set to 512 and 1024, the test score decreased, and both test accuracy and validation accuracy increased. This demonstrates that increasing the number of neurons allows the model to better learn and generalize from the data. I also noticed that increasing neurons from 128 to 1024 gives only a small improvement because the model already learns most patterns from the training samples, so extra neurons for processing doen't improve performance.